xds/src/main/java/io/grpc/xds/client/XdsResourceType.java

```diff
@@ -148,15 +148,33 @@ ValidatedResourceUpdate<T> parse(Args args, List<Any> resources) {
       Any resource = resources.get(i);
 
       Message unpackedMessage;
+      String wrappedResourceName = null;
       try {
-        resource = maybeUnwrapResources(resource);
+        if (resource.getTypeUrl().equals(TYPE_URL_RESOURCE)) {
+          Resource wrappedResource = unpackCompatibleType(resource, Resource.class,
+              TYPE_URL_RESOURCE, null);
+          resource = wrappedResource.getResource();
+          // Check for resource name in Name of the wrapped Resource, if not also check
+          // the ResourceName field.
+          if (wrappedResource.getName().isEmpty()) {
+            wrappedResourceName = wrappedResource.hasResourceName()
+                ? wrappedResource.getResourceName().getName()
+                : null;
+          } else {
+            wrappedResourceName = wrappedResource.getName();
+          }
+        } 
         unpackedMessage = unpackCompatibleType(resource, unpackedClassName(), typeUrl(), null);
       } catch (InvalidProtocolBufferException e) {
         errors.add(String.format("%s response Resource index %d - can't decode %s: %s",
                 typeName(), i, unpackedClassName().getSimpleName(), e.getMessage()));
         continue;
       }
       String name = extractResourceName(unpackedMessage);
+      // Fallback to wrapped resource name, if the inner resource finally didn't have a name.
+      if (name.isEmpty()) {
```

Comment:
Why is this checking `isEmpty()` instead of `null`? I thought the problem was the unpackedMessage had no name field.

Reply:
Fixed, you are right it should be null not emtpy.